ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala

```diff
@@ -0,0 +1,149 @@
+  NotPruned,
+  Offset,
+  ParticipantPruned,
+  SubmissionId,
+  WriteParticipantPruningService
+}
+import com.daml.lf.data.Ref
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.api.grpc.GrpcApiService
+import com.daml.platform.server.api.validation.ErrorFactories
+import com.google.protobuf.empty.Empty
+import io.grpc.{ServerServiceDefinition, StatusRuntimeException}
+
+import scala.compat.java8.FutureConverters
+import scala.concurrent.{ExecutionContext, Future}
+
+final class ApiParticipantPruningService private (
+    readBackend: IndexParticipantPruningService with LedgerEndService,
+    writeBackend: WriteParticipantPruningService)(
+    implicit grpcExecutionContext: ExecutionContext,
+    logCtx: LoggingContext)
+    extends ParticipantPruningServiceGrpc.ParticipantPruningService
+    with GrpcApiService {
+
+  private val logger = ContextualizedLogger.get(this.getClass)
+
+  override def bindService(): ServerServiceDefinition =
+    ParticipantPruningServiceGrpc.bindService(this, DirectExecutionContext)
+
+  override def prune(request: PruneRequest): Future[Empty] = {
+    val submissionIdOrErr = SubmissionId
+      .fromString(
+        if (request.submissionId.nonEmpty) request.submissionId else UUID.randomUUID().toString)
+      .left
+      .map(err => ErrorFactories.invalidArgument(s"submission_id $err"))
+
+    submissionIdOrErr.fold(
+      Future.failed,
+      submissionId =>
+        LoggingContext.withEnrichedLoggingContext("submissionId" -> submissionId) {
+          implicit logCtx =>
+            (for {
+
+              pruneUpTo <- validateRequest(request: PruneRequest)
+
+              _ <- pruneWriteService(pruneUpTo, submissionId)
+
+              pruneResponse <- pruneLedgerApiServerIndex(pruneUpTo)
+
+            } yield pruneResponse).andThen(logger.logErrorsOnCall[Empty])
+      }
+    )
+  }
+
+  private def validateRequest(request: PruneRequest)(
+      implicit logCtx: LoggingContext): Future[Offset] = {
+    (for {
+      pruneUpToProto <- checkOffsetIsSpecified(request.pruneUpTo)
+      pruneUpToString <- checkOffsetIsAbsolute(pruneUpToProto)
+      pruneUpTo <- checkOffsetIsHexadecimal(pruneUpToString)
+    } yield (pruneUpTo, pruneUpToString))
+      .fold(Future.failed, o => checkOffsetIsBeforeLedgerEnd(o._1, o._2))
+  }
+
+  private def pruneWriteService(pruneUpTo: Offset, submissionId: SubmissionId)(
+      implicit logCtx: LoggingContext): Future[Unit] =
+    FutureConverters
+      .toScala(writeBackend.prune(pruneUpTo, submissionId))
+      .flatMap {
+        case NotPruned(status) => Future.failed(ErrorFactories.grpcError(status))
+        case ParticipantPruned =>
+          logger.info(s"Pruned participant ledger up to ${pruneUpTo.toString} inclusively.")
+          Future.successful(())
+      }
+
+  private def pruneLedgerApiServerIndex(pruneUpTo: Offset)(implicit logCtx: LoggingContext) =
+    readBackend
+      .prune(pruneUpTo)
+      .map { _ =>
```

Comment:
Same here, print a log line if the pruning fails.

Reply:
See above.